catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,434 @@
 
-    this.runner = incomingRunner;
+      boolean sourceIsNotInCache = cache.getIfPresent(sourceKey) == null;
+      if (sourceIsNotInCache) {
+        LOGGER.debug("Found a new SourceKey for source id={}", source.getId());
+        cache.put(sourceKey, SourceAvailability.unknown());
+      }
 
-    scheduler =
-        Executors.newSingleThreadScheduledExecutor(
-            StandardThreadFactoryBuilder.newThreadFactory("sourcePollerThread"));
+      LOGGER.trace("Starting a process to check the availability of source id={}", source.getId());
+      Future<SourceAvailability> sourceAvailabilityFuture =
+          Futures.withTimeout(
+              sourceAvailabilityChecksThreadPool.submit(() -> checkAvailability(source)),
+              DEFAULT_TIMEOUT,
+              DEFAULT_TIMEOUT_TIME_UNIT,
+              pollAllSourcesThreadPool);
+      availabilityFutures.put(sourceKey, sourceAvailabilityFuture);
+    }
+    LOGGER.trace("Successfully submitted availability checks for all of the sources");
+    return availabilityFutures;
+  }
+
+  private void CacheResults(Map<SourceKey, Future<SourceAvailability>> availabilityFutures) {
+    for (final Entry<SourceKey, Future<SourceAvailability>> entry :
+        availabilityFutures.entrySet()) {
+      final Future<SourceAvailability> future = entry.getValue();
+      final SourceKey sourceKey = entry.getKey();
+      final String sourceId = sourceKey.id;
+
+      try {
+        SourceAvailability result = future.get();
+        cache.put(sourceKey, result);
+      } catch (CancellationException e) {
+        future.cancel(true);
+        LOGGER.debug("Unable to check the availability of source id={}", sourceId, e);
+      } catch (RuntimeException | ExecutionException e) {
+        future.cancel(true);
+
+        if (e.getCause() instanceof TimeoutException) {
+          LOGGER.debug(
+              "Timeout occurred while checking the availability of source id={}", sourceId);
+          cache.put(sourceKey, new SourceAvailability(SourceStatus.TIMEOUT));
+        } else {
+          LOGGER.debug("Exception checking the availability of source id={}", sourceId, e);
+          cache.put(sourceKey, new SourceAvailability(SourceStatus.EXCEPTION));
+        }
+      } catch (InterruptedException e) {
+        future.cancel(true);
+        LOGGER.debug("Interrupted while checking the availability of source id={}", sourceId);
+        Thread.currentThread().interrupt();
+      }
+    }
+    LOGGER.trace("Cache updated with new source availabilities");
+  }
+
+  private SourceAvailability checkAvailability(final Source source) {
+    SourceStatus newSourceStatus =
+        source.isAvailable() ? SourceStatus.AVAILABLE : SourceStatus.UNAVAILABLE;
+    return new SourceAvailability(newSourceStatus);
+  }
 
-    handle = scheduler.scheduleAtFixedRate(runner, INITIAL_DELAY, interval, TimeUnit.MINUTES);
+  private void setPollInterval(final long pollInterval, final TimeUnit pollIntervalTimeUnit) {
```

Comment:
I think that would be a good idea

Reply:
I assume that means removing the pollIntervalTimeUnit variable and just using the constant? Should I also just use the default of seconds for the timout time unit? This would reduce the number of parameters for our methods.

Reply:
Actually when testing I want to be able to make these values really small so maybe we should keep this method around?

Reply:
Since I'm mocking out the ExecutorServices in the tests now, the poll interval no longer affects the tests. Maybe we should just have setPollIntervalMinutes

Reply:
`setPollIntervalMinutes` doesn't do anything right now. It needs to re-do the `scheduleAtFixedRate` part, right?